### PR TITLE
Migrate the core functionality from the remote utility to aexpect

### PR DESCRIPTION
**Pull request by pevogam**
This code makes more sense in the aexpect repo as indicated by the
higher coupling of imports and use cases.

This commit goes hand in hand with the relevant commits on the aexpect
side importing the raw code and extending on exceptions, log functions
and so on.

Signed-off-by: Plamen Dimitrov <pdimitrov@pevogam.com>

### CODE REVIEW

**Comment by pevogam** [comment]
This pull request goes together with [PR #55 on aexpect](https://github.com/avocado-framework/aexpect/pull/55) and will require a very significant testing for breakage.

**Comment by pevogam** [comment]
I have ran 700+ tests from our own test suite using avocado-vt on this branch (using aexpect migrated `remote` module) and encountered no errors so the branch works to the best of my knowledge and valiadtion. Regarding the two linters:

1) Travis CI fails with identical error on all instances which is

```
E0611: xx,0: : No name 'remote' in module 'aexpect'
```

and is due to the fact that the PR on the side of `aexpect` hasn't been merged yet. Depending on the deployment technique used by Travis we might also need an official release on the aexpect side.

2) LGTM complains about double import statement but this is the cleanest way to use remote from aexpect with a minimum of modifications on the avocado-vt side.

Any advice on these two and a review is very welcome.

File: virttest/ssh_key.py
Comment (balamuruhans) [inline virttest/ssh_key.py]: why don't we consider `RemoteFile()` and `RemoteRunner()` to be moved to aexpect ?

File: virttest/virt_vm.py
Comment (balamuruhans) [inline virttest/virt_vm.py]: `remote_commander()` as well ?

File: virttest/ssh_key.py
Comment (pevogam) [reply virttest/ssh_key.py]: Hi @balamuruhans, these:

1) involve additional dependencies that are specific to avocado-vt
2) are rarely used and extra complicated
3) duplicate most of the core functionality already provided by remote and thus aexpect

For more details and sources for all these three conclusions, please see [the analogous discussion on the aexpect side](https://github.com/avocado-framework/aexpect/pull/55#issuecomment-493401973).

File: virttest/virt_vm.py
Comment (pevogam) [reply virttest/virt_vm.py]: Same reasoning as the one above.

**Comment by pevogam** [comment]
The current branch now has ever increasing number of conflitcs with the current master. @ldoktor Can we speed up the process of reviewing/merging on the aexpect side? I have already proposed three or four ways to remedy the process there for the last 5 months and if you don't like any of them please mention it so that we could make decision on whether to cancel this migration. A PR of this side requires greater attention since it does a large number of changes and if left unattended for long becomes vulnerable to more and more conflicts.

**Comment by pevogam** [comment]
Hi @ldoktor, @clebergnu, @balamuruhans, all current checks fail because of the missing aexpect dependency. I am not sure how to remedy this and have no control over it so I am wondering what should we do next. The `remote.py` code is now fully merged to aexpect master but I guess not packaged or something which is why it is currently missing in all these checks.

Regarding rebase conflicts, the only troublesome rebase I had to do is with 8c26befad09f7093be205edcad3508c3899a38df which adds an extra argument to `remote_login(..., user_known_hosts_file='/dev/null')`. It can be fixed rather easily and there are no other conflicts that I am aware of so if you know such please let me know and let's register them here.

**Comment by pevogam** [comment]
Hi @ldoktor,

> now it's time to polish the Avocado-vt part. I can see 2 options:
> 
>     1. bump the requirements to the new aexpect that contains remote commander
> 
>     2. do try/except imports of remote commander for a transition period until everything settles.
> 
> 
> Note there is no aexpect with remote commander just yet, but I already pinged Lucas to release it.

I guess the plan if following option 1 then is to wait until the release by Lucas, then bump the aexpect version in the (normal and travis) requirements. Not directly related to this but I hope https://github.com/avocado-framework/aexpect/pull/57 could make it into the release too.

> Until them please consider the cost of allowing the use of old aexpect, but after a brief look at this it should be fairly simple, just use:
> 
> ```
> try:
>     from aexpect import remote
> except ImportError:
>     from virttest import remote
> ```
> 
> and when new aexpect is installed, it should use that. When everything settles, we could remove it and bump the aexpect version together with removing the remote commander code from virttest. What do you think?

I don't think option 2. makes much sense since this PR will also remove functionality from the `virttest/remote` module. If we catch the missing aexpect dependency and import the local already modified module, this will break things for all people not having the dependency either way. It is better to break it through import and thus point directly to the issue instead of letting them deal with missing functions because we imported the old module that now only contains code used by a few people we didn't want to migrate. The moment this PR is merged the `virttest/remote` module will become the minimal form so changing the imports above will only make sense if done in advance.

> Anyway both solutions are fine by me so if you trust the aexpect will work for everyone, feel free to suggest the new aexpect in requirements immediately (when it's ready).

I will be available to fix anything that comes up within hours or so but I don't expect such problems. There are a few situations of later merged code but we can only account for them (fast hopefully) once this PR is merged. Any subsequent merges conflicting with this large migration have this effect which is why I iterated before that there will be less damage if migrated quickly.

**Comment by ldoktor** [comment]
I do agree about the framework part, but there are also the `tp-*` repos. We can't do the hard switch there as people still need to be able to contribute and update tp- even when they use LTS, so until there is an LTS with dependency on aexpect with remoter class we need to either modify all tp- usage to use aexpect or avocado-vt, or we need to re-import the removed objects in `virttest.remote` module. I'd prefer the second as we already have similar workarounds in `virttest/utils_misc.py` (reimport `avocado.utils.process` methods) and it requires less changes to tests and tests are something that is not always upstream...

As for the scale on slightly outdated `tp-*`

    fgrep "from virttest import remote" * -r | wc -l
    91

So in `virttest.remote` we should add:

```
# TODO: Remove when Avocado-69.0 LTS becomes unsupported
from aexpect import LoginError, LoginAuthenticationError, LoginTimeoutError, ...
```

That way the old Avocado that depends on old Aexpect still contains the `virttest.remote.LoginError` and the new Avocado that depends on new Aexpect also provides the `virttest.remote.LoginError` that is the same as `aexpect.remote.LoginError` so tp- code will still work on both.

Anyway I tested this briefly, it seems to work so let's just wait for the new aexpect, rerun the CI and see how it stands there.

**Comment by pevogam** [comment]
There were new conflicts with the 79.0 master branch so I had to resolve them and push again. @lmr Is there a chance we could release the new aexpect version soon so that we could proceed with this? So far I had to resolve conflicts with constantly changing base 3-4 times, some easier and some harder.

**Comment by pevogam** [comment]
> There were new conflicts with the 79.0 master branch so I had to resolve them and push again. @lmr Is there a chance we could release the new aexpect version soon so that we could proceed with this? So far I had to resolve conflicts with constantly changing base 3-4 times, some easier and some harder.

I see all aexpect PRs needed for its release are in place, thanks @lmr. I suggest using `1.6.0` for its release tag which is the final step together with enabling the dependency here so that I can proceed with this.

**Comment by pevogam** [comment]
Hi all, after a recent conflict that modifies the already migrated `remote.py` (the devs there cannot know it was migrated by any means), with growing changes that might not take effect on the aexpect side after this PR was proposed, after fixing three of four more collected conflicts in the last months, and despite numerous warnings on my side that this has to happen at least a bit faster but it keeps taking months after months, I must confess I am getting tired of this.

I did this migration for the benefit of the Avocado VT repo and have even introduces a few extra session classes on the aexpect side, did some exception handling refactoring, and more. If the maintainers of this repo are not willing to take it, they should say so. If they don't have the time for it, they should give control to others willing to drive the repos forward. I see no other way around this.

@lmr As you seem to be busy with other projects I would suggest that you increase the number of maintainers of the aexpect project. I understand your lack of time and resources and I don't find anything wrong with it. What I find wrong is blocking the progress and creating artificial labor for other contributors that want to improve it. I have spent and keep spending time adjusting this migration and my concern with potential breakage only grows, considering this PR was created many months ago when everything was in place and for months I keep collecting cases of people doing changes on the `remote.py` module.

@ldoktor I think we should not have embarked on this since I already had to block a PR on the avocado side where a contributed started reinventing the wheel and writing their own aexpect utility within avocado. I pointed them to importing aexpect and their feedback on the repo was not very ensuring (https://github.com/avocado-framework/avocado/pull/3669#discussion_r404318854). I found it odd that I had to defend a project that I myself had to spend a lot of time waiting to contribute on, all for the sake of modularity.

To both, I hope you understand how I feel about this - having to defend the almost absent maintenance of aexpect on one side and having to keep doing unnecessary extra efforts here and sending reminders. I don't claim any of your resources, I only say that if you don't have them you should let more people in, the ones willing to take care of these projects.

**Comment by sathnaga** [comment]
Travis Failure can be ignore as it will get fixed once aexpect new release is done and we change the dependency to install that new release.

Tested bootstrap with latest avocado-vt + this patch + aexpect master, it works fine..

with latest expect:
```
# avocado vt-bootstrap 
Running bootstrap for qemu

1 - Checking the mandatory programs and headers
/bin/xz OK
/sbin/tcpdump OK
/bin/nc OK
/sbin/ip OK
/bin/arping OK
/bin/gcc OK
/usr/include/sys/unistd.h OK
/usr/include/sys/types.h OK
/usr/include/sys/socket.h OK
/usr/include/python2.7/Python.h OK

...
```

without latest aexpect:
```
# avocado run boot --vt-type qemu 
Failed to load plugin from module "avocado_vt.plugins.vt": ImportError("cannot import name 'remote' from 'aexpect' (/usr/local/lib/python3.7/site-packages/aexpect/__init__.py)") :
  File "/usr/local/lib/python3.7/site-packages/avocado_framework-80.0-py3.7.egg/avocado/core/extension_manager.py", line 62, in __init__
    plugin = ep.load()
  File "/usr/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2443, in load
    return self.resolve()
  File "/usr/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2449, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/local/lib/python3.7/site-packages/avocado_framework_plugin_vt-80.0-py3.7.egg/avocado_vt/plugins/vt.py", line 32, in <module>
    from ..loader import VirtTestLoader
  File "/usr/local/lib/python3.7/site-packages/avocado_framework_plugin_vt-80.0-py3.7.egg/avocado_vt/loader.py", line 29, in <module>
    from virttest import storage
  File "/usr/local/lib/python3.7/site-packages/avocado_framework_plugin_vt-80.0-py3.7.egg/virttest/storage.py", line 24, in <module>
    from virttest import iscsi
  File "/usr/local/lib/python3.7/site-packages/avocado_framework_plugin_vt-80.0-py3.7.egg/virttest/iscsi.py", line 22, in <module>
    from virttest import utils_net
  File "/usr/local/lib/python3.7/site-packages/avocado_framework_plugin_vt-80.0-py3.7.egg/virttest/utils_net.py", line 21, in <module>
    from aexpect import remote
```


So this demands, the aexpect tagging and release ASAP and  change of https://github.com/avocado-framework/avocado-vt/blob/master/requirements.txt#L2 and https://github.com/avocado-framework/avocado-vt/blob/master/requirements-travis.txt#L7
and I guess update to pypi(https://pypi.org/search/?q=aexpect) also needed, probably release would take care of it .
@ldoktor @lmr needs your help here. Thanks :+1:

**Review by sathnaga — APPROVED** [review]
with sanity testing, LGTM, it good to get this merge sooner, especially as it involves lot of file dependencies and re-basing would be difficult task, as pre-requisite we need help from @ldoktor @lmr to release aexpect module.

I hope @pevogam will be happy to help incase of any regressions/hotfixes.

**Comment by pevogam** [comment]
Thanks @sathnaga for your assistance! The blocker here remains the release though and less the review part for which I know we could go faster. The ultimate blocker remains the aexpect project in fact.

**Comment by clebergnu** [comment]
@pevogam by release you mean commit access to the aexpect repo and to PyPI right? I don't see why access couldn't/shouldn't be arranged for you, if you're willing to (also) do releases.  Would you be willing to?

**Comment by pevogam** [comment]
> @pevogam by release you mean commit access to the aexpect repo and to PyPI right? I don't see why access couldn't/shouldn't be arranged for you, if you're willing to (also) do releases. Would you be willing to?

Hi @clebergnu, thanks a lot for chiming in! My slightly irritated (sorry for that) paragraph above is talking in general about balancing the load that maintainers of these repos have. But clearly my words also consider me as an instance of someone that could help out as long as you need this help. So if you also suggest adding more resources to maintain a repo like aexpect as a solution, I don't mind helping with whatever I can.

**Comment by pevogam** [comment]
And while I was waiting for more replies here, even more conflicts were collected:

#2720 
#2703 

The first one introduces entire classes in the previously migrated module...

**Comment by sathnaga** [comment]
@clebergnu @ldoktor what's next here?

**Comment by sathnaga** [comment]
this would be very good candidate for 82 LTS.

**Comment by pevogam** [comment]
> this would be very good candidate for 82 LTS.

I guess not for LTS since a lot of damage was done on this PR and will need some brave follow-up fixing but I am thinking of dropping the PR if it doesn't get merged soon. I simply cannot stay and watch a pile of work and hotfixes growing on my desk while not being able to do anything about it. If it doesn't move anywhere then it is probably of insufficient interest to the project maintainers even though I was originally asked and convinced by a maintainer of the project to do it for everyone.

**Comment by sathnaga** [comment]
> > this would be very good candidate for 82 LTS.
> 
> I guess not for LTS since a lot of damage was done on this PR and will need some brave follow-up fixing but I am thinking of dropping the PR if it doesn't get merged soon. I simply cannot stay and watch a pile of work and hotfixes growing on my desk while not being able to do anything about it. If it doesn't move anywhere then it is probably of insufficient interest to the project maintainers even though I was originally asked and convinced by a maintainer of the project to do it for everyone.

@pevogam I truly understand your concern and share both side of the coin being a maintainer, please do not loose the hope,
this PR just need a second ACK to get this merged for Avocado-VT, I was more inclined towards getting it merged for 82LTS, given that , it was not your usecase, we are good, I agree your pain on rebasing, hopefully, let this be your last rebase and fixing conflicts and we will get this merged. Appreciate your valued contributions.

Having said that aexpect repo side needs some activity, let try to fasten it aswell.

**Comment by pevogam** [comment]
Hi @sathnaga, I have resolved any current conflicts (with some fixes to add once this is part of master) but it seems the CI is not functional in general. What should we do about its checks?

**Comment by pevogam** [comment]
> This pull request **introduces 19 alerts** and **fixes 4** when merging [3e3a6e4](https://github.com/avocado-framework/avocado-vt/commit/3e3a6e45818d9ae07ea430684e69eb9ca58604d0) into [331a349](https://github.com/avocado-framework/avocado-vt/commit/331a3492c3b1160f54b554b0d9260b6c8b0d9e1f) - [view on LGTM.com](https://lgtm.com/projects/g/avocado-framework/avocado-vt/rev/pr-f3279418a4d81bbdec2bc81581b6afe101dd5170)
> 
> **new alerts:**
> 
>     * 16 for Module is imported with 'import' and 'import from'
> 
>     * 2 for Wrong name for an argument in a call
> 
>     * 1 for Unused import
> 
> 
> **fixed alerts:**
> 
>     * 2 for Clear-text logging of sensitive information
> 
>     * 1 for Should use a 'with' statement
> 
>     * 1 for Wrong number of arguments in a class instantiation

Most complaints are of the form "Module 'aexpect' is imported with both 'import' and 'import from'" which is due to the fact that we minimize line count of the diff while only reimporting aexpect's remote submodule. As we can see from the number of conflicts this had to go through this was no doubt a good idea. Once this migration is complete and some follow up PRs add fixes on possible breakage from the past year, we could also afford more module renames and fixing these LGTM problems.

**Comment by pevogam** [comment]
Hi @sathnaga, I hope we can speed this up before more conflicts emerge. Thanks!

**Comment by sathnaga** [comment]
@pevogam thanks for the patience, sorry, I was out for a while, I guess we are in final stages on getting this merged.

I was checking on the travis failure and noticing still travis is using, older axpect and hence failing, probably you have to update the requirements-travis.txt as part of this PR.
```
Collecting aexpect==1.5.1 (from -r requirements-travis.txt (line 7))
```

**Comment by pevogam** [comment]
> @pevogam thanks for the patience, sorry, I was out for a while, I guess we are in final stages on getting this merged.
> 
> I was checking on the travis failure and noticing still travis is using, older axpect and hence failing, probably you have to update the requirements-travis.txt as part of this PR.
> 
> ```
> Collecting aexpect==1.5.1 (from -r requirements-travis.txt (line 7))
> ```

Alright, I fixed all additional CI errors from conflicts in the last months and the CI is green now. The biggest regression so far is in having to add a temporary `remote_login` function to handle some additional SSH options that weren't present at the time of the migration, see diff for details.

In addition, I have created an issue on the aexpect side https://github.com/avocado-framework/aexpect/issues/69#issue-721321876 for tracking any problems and need for hotfixes arising from merging this PR. I really hope this can finally be merged now so that we can follow up on any fixes rather than getting more conflicts.

@sathnaga Can you recommend an active maintainer that is willing to add a second review here?

**Review by sathnaga — APPROVED** [review]
LGTM

**Comment by sathnaga** [comment]
@luckyh @chunfuwen @dzhengfy  can one you be second reviewer for this, this is pending from long and @pevogam has done a lot of rebasing already and he is willing to hotfix incase of any hiccups/regressions aswel.

Thanks in advance.

**Review by ldoktor — COMMENTED** [review]
Hello Plamen, in general it looks good although I don't like much the rename to `remote_old`, rather I'd use `aremote` for aexpect remote, or `vremote` for virttest remote. If you consider this as a small step towards removing `remote` completely from virttest than I'm fine to call it `remote_old` for the transition period, but if you don't intend to move all remote stuff from virttest then I'd suggest renaming it (`aremote`, `vremote` or other name that would represent the module).

Another thing is the use of `remote` in tp-qemu, tp-libvirt and tp-spice`. Is there a connected PR to merge tp-* changes together? If not than the `virttest.remote` must keep all the interfaces in-place for the transition period. If the API hadn't change I'd recommended using `from aexpect.remote import *` in `virttest.remote` or adding simple wrappers for each method that is being used:

```
$ fgrep "remote." * -r  | wc -l
224
$ fgrep "from virttest import remote" * -r | wc -l
92
$ fgrep "remote." * -r
io-github-autotest-libvirt/libvirt/tests/src/nwfilter/nwfilter_edit_uuid.py:            remote.handle_prompts(session, None, None, r"[\#\$]\s*$")
io-github-autotest-libvirt/libvirt/tests/src/nwfilter/nwfilter_edit_uuid.py:        except (aexpect.ShellError, aexpect.ExpectError, remote.LoginTimeoutError) as details:
io-github-autotest-libvirt/libvirt/tests/src/multiqueue.py:    host_session = remote.remote_login("ssh", server_ip, 22, "root",
io-github-autotest-libvirt/libvirt/tests/src/libvirt_hugepage.py:                remote.scp_to_remote(vm.get_address(), 22,
io-github-autotest-libvirt/libvirt/tests/src/macvtap.py:        remote_file = remote.RemoteFile(vm.get_address(), 'scp', 'root',
io-github-autotest-libvirt/libvirt/tests/src/macvtap.py:        except remote.LoginTimeoutError as fail:
io-github-autotest-libvirt/libvirt/tests/src/macvtap.py:            except remote.LoginTimeoutError as fail:
io-github-autotest-libvirt/libvirt/tests/src/daemon/restart_consist.py:    except remote.LoginError:
...
```

**Comment by pevogam** [comment]
Hi Lukas,

> Hello Plamen, in general it looks good although I don't like much the rename to `remote_old`, rather I'd use `aremote` for aexpect remote, or `vremote` for virttest remote. If you consider this as a small step towards removing `remote` completely from virttest than I'm fine to call it `remote_old` for the transition period, but if you don't intend to move all remote stuff from virttest then I'd suggest renaming it (`aremote`, `vremote` or other name that would represent the module).

I did this as a step towards deprecating the rest of the functionality here yes, leaving a few classes that duplicated functionality IMO and after some research here turned out nobody was using. I kept them around just in case and decided to call this `remote_old` with the intent of eventually removing it if no one complains.

However, as so much time passed, I think right now some people started using some of these classes (`Remote_Package`, `RemoteFile`, `RemoteRunner`, etc.) for unknown to me reasons. So the end intend is to keep the local module, however try to stimulate people to either 1) move away from it using better tools or 2) repurpose the module to something new and unique that doesn't intersect with `aexpect`. I don't think the `aremote` and `vremote` solved anything at the present since it doesn't achieve this goal by any means. The plan is rather to see what is left in the module and how much it is used and then find a proper name for the local `remote.py` module. There is a similar problem with the multiple remote imports (LGTM detections above) that has to be addressed in the same round. For the time being however, the current name will dissuade people from relying too much on this deprecated functionality or find a new meaning (and name) for it.

You can also see some of the `_to_migrate` methods I have re-added for the same reason.

> Another thing is the use of `remote` in tp-qemu, tp-libvirt and tp-spice. Is there a connected PR to merge tp-* changes together?

This is why I so much needed a maintainer to help out with this and provide such hints from the start. There are no such PR-s there.

> If not than the `virttest.remote` must keep all the interfaces in-place for the transition period. If the API hadn't change I'd recommended using `from aexpect.remote import *` in `virttest.remote` or adding simple wrappers for each method that is being used:
> 
> ```
> $ fgrep "remote." * -r  | wc -l
> 224
> $ fgrep "from virttest import remote" * -r | wc -l
> 92
> $ fgrep "remote." * -r
> io-github-autotest-libvirt/libvirt/tests/src/nwfilter/nwfilter_edit_uuid.py:            remote.handle_prompts(session, None, None, r"[\#\$]\s*$")
> io-github-autotest-libvirt/libvirt/tests/src/nwfilter/nwfilter_edit_uuid.py:        except (aexpect.ShellError, aexpect.ExpectError, remote.LoginTimeoutError) as details:
> io-github-autotest-libvirt/libvirt/tests/src/multiqueue.py:    host_session = remote.remote_login("ssh", server_ip, 22, "root",
> io-github-autotest-libvirt/libvirt/tests/src/libvirt_hugepage.py:                remote.scp_to_remote(vm.get_address(), 22,
> io-github-autotest-libvirt/libvirt/tests/src/macvtap.py:        remote_file = remote.RemoteFile(vm.get_address(), 'scp', 'root',
> io-github-autotest-libvirt/libvirt/tests/src/macvtap.py:        except remote.LoginTimeoutError as fail:
> io-github-autotest-libvirt/libvirt/tests/src/macvtap.py:            except remote.LoginTimeoutError as fail:
> io-github-autotest-libvirt/libvirt/tests/src/daemon/restart_consist.py:    except remote.LoginError:
> ...
> ```

I don't think the API has changed other than the few methods I had to re-insert back as a regression from the long time but I cannot fully guarantee there won't be some fires to quickly put out in these repos. I would go for the `from aexpect.remote import *` approach into the `virttest.remote` namespace and this is the most I can do as wrapping each of tens of methods manually is not feasible at this stage.

**Comment by ldoktor** [comment]
> Hi Lukas,
> 
> > Hello Plamen, in general it looks good although I don't like much the rename to `remote_old`, rather I'd use `aremote` for aexpect remote, or `vremote` for virttest remote. If you consider this as a small step towards removing `remote` completely from virttest than I'm fine to call it `remote_old` for the transition period, but if you don't intend to move all remote stuff from virttest then I'd suggest renaming it (`aremote`, `vremote` or other name that would represent the module).
> 
> I did this as a step towards deprecating the rest of the functionality here yes, leaving a few classes that duplicated functionality IMO and after some research here turned out nobody was using. I kept them around just in case and decided to call this `remote_old` with the intent of eventually removing it if no one complains.
> 
> However, as so much time passed, I think right now some people started using some of these classes (`Remote_Package`, `RemoteFile`, `RemoteRunner`, etc.) for unknown to me reasons. So the end intend is to keep the local module, however try to stimulate people to either 1) move away from it using better tools or 2) repurpose the module to something new and unique that doesn't intersect with `aexpect`. I don't think the `aremote` and `vremote` solved anything at the present since it doesn't achieve this goal by any means. The plan is rather to see what is left in the module and how much it is used and then find a proper name for the local `remote.py` module. There is a similar problem with the multiple remote imports (LGTM detections above) that has to be addressed in the same round. For the time being however, the current name will dissuade people from relying too much on this deprecated functionality or find a new meaning (and name) for it.
> 
Sure I do understand that and `remote_old` is good-enough-name for now, but we should think of a better naming when this whole transition is over. I think the `vremote` name would suit a bit better, but it's not necessary to do now.

> You can also see some of the `_to_migrate` methods I have re-added for the same reason.
> 
> > Another thing is the use of `remote` in tp-qemu, tp-libvirt and tp-spice. Is there a connected PR to merge tp-* changes together?
> 
> This is why I so much needed a maintainer to help out with this and provide such hints from the start. There are no such PR-s there.

Yeah, I'm sorry for that and btw officially I should not work on Avocado-vt (for a few years already, but it's hard to just stop).

> 
> > If not than the `virttest.remote` must keep all the interfaces in-place for the transition period. If the API hadn't change I'd recommended using `from aexpect.remote import *` in `virttest.remote` or adding simple wrappers for each method that is being used:
> > ```
> > $ fgrep "remote." * -r  | wc -l
> > 224
> > $ fgrep "from virttest import remote" * -r | wc -l
> > 92
> > $ fgrep "remote." * -r
> > io-github-autotest-libvirt/libvirt/tests/src/nwfilter/nwfilter_edit_uuid.py:            remote.handle_prompts(session, None, None, r"[\#\$]\s*$")
> > io-github-autotest-libvirt/libvirt/tests/src/nwfilter/nwfilter_edit_uuid.py:        except (aexpect.ShellError, aexpect.ExpectError, remote.LoginTimeoutError) as details:
> > io-github-autotest-libvirt/libvirt/tests/src/multiqueue.py:    host_session = remote.remote_login("ssh", server_ip, 22, "root",
> > io-github-autotest-libvirt/libvirt/tests/src/libvirt_hugepage.py:                remote.scp_to_remote(vm.get_address(), 22,
> > io-github-autotest-libvirt/libvirt/tests/src/macvtap.py:        remote_file = remote.RemoteFile(vm.get_address(), 'scp', 'root',
> > io-github-autotest-libvirt/libvirt/tests/src/macvtap.py:        except remote.LoginTimeoutError as fail:
> > io-github-autotest-libvirt/libvirt/tests/src/macvtap.py:            except remote.LoginTimeoutError as fail:
> > io-github-autotest-libvirt/libvirt/tests/src/daemon/restart_consist.py:    except remote.LoginError:
> > ...
> > ```
> 
> I don't think the API has changed other than the few methods I had to re-insert back as a regression from the long time but I cannot fully guarantee there won't be some fires to quickly put out in these repos. I would go for the `from aexpect.remote import *` approach into the `virttest.remote` namespace and this is the most I can do as wrapping each of tens of methods manually is not feasible at this stage.

Sure, which is why I suggested it. I tried some of the APIs and they look alike so let's get this in and hammer-out the potential issues if they pop-up. Overall this is a big step into the right direction.

**Review by ldoktor — APPROVED** [review]
OK, it's hard but the few APIs I tested were alike so the tests using `virttest.remote` should hopefully keep working the same. There might be some issues here and there so hopefully people will address these either in aexpect or in a transition layer in `virttest.remote`.

**Comment by pevogam** [comment]
Thanks @sathnaga and @ldoktor! I will be around the entire day today and keep vigil in case we get some urgent issues, otherwise I will remain around in the normal way if somebody from the VT users needs further assistance with this.

**Comment by zhenyzha** [comment]
@pevogam hello This patch causes the following error to be reported when tp-qemu is running:
avocado run: error: unrecognized arguments: --vt-config
Perhaps a plugin is missing; run 'avocado plugins' to list the installed ones

**Comment by pevogam** [comment]
Hi @zhenyzha,

> @pevogam hello This patch causes the following error to be reported when tp-qemu is running:
> 
> ```
> avocado run: error: unrecognized arguments: --vt-config
> Perhaps a plugin is missing; run 'avocado plugins' to list the installed ones
> ```

This is odd as I wouldn't expect arguments on the command line to be affected in any way. Do you have a link to the CI logs or other output logs?

**Comment by zhenyzha** [comment]
> This is odd as I wouldn't expect arguments on the command line to be affected in any way. Do you have a link to the CI logs or other output logs?
Hi @pevogam 

Sorry for the delay.

`VT-BOOTSTRAP FINISHED
You may take a look at the following online docs for more info:
 - http://avocado-vt.readthedocs.org/
 - http://avocado-framework.readthedocs.org/
INFO | Creating test configuration file '/var/lib/avocado/data/avocado-vt/backends/qemu/cfg/tests-example.cfg'
INFO | Assigning vcpu num '64'
INFO | Assigning memory size '21504'
INFO | Start to run test
Failed to load plugin from module "avocado_vt.plugins.vt": ModuleNotFoundError("No module named 'aexpect.remote'",) :
  File "/home/kar/workspace/avocado/avocado/core/extension_manager.py", line 62, in __init__
    plugin = ep.load()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2318, in load
    return self.resolve()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2324, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/kar/workspace/avocado-vt/avocado_vt/plugins/vt.py", line 32, in <module>
    from ..loader import VirtTestLoader
  File "/home/kar/workspace/avocado-vt/avocado_vt/loader.py", line 29, in <module>
    from virttest import storage
  File "/home/kar/workspace/avocado-vt/virttest/storage.py", line 22, in <module>
    from virttest import nbd
  File "/home/kar/workspace/avocado-vt/virttest/nbd.py", line 5, in <module>
    from virttest import utils_misc
  File "/home/kar/workspace/avocado-vt/virttest/utils_misc.py", line 83, in <module>
    from virttest import utils_disk
  File "/home/kar/workspace/avocado-vt/virttest/utils_disk.py", line 27, in <module>
    from virttest import remote
  File "/home/kar/workspace/avocado-vt/virttest/remote.py", line 13, in <module>
    from aexpect.remote import *

Failed to load plugin from module "avocado_vt.plugins.vt_list": ModuleNotFoundError("No module named 'aexpect.remote'",) :
  File "/home/kar/workspace/avocado/avocado/core/extension_manager.py", line 62, in __init__
    plugin = ep.load()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2318, in load
    return self.resolve()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2324, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/kar/workspace/avocado-vt/avocado_vt/plugins/vt_list.py", line 26, in <module>
    from .vt import add_basic_vt_options, add_qemu_bin_vt_option
  File "/home/kar/workspace/avocado-vt/avocado_vt/plugins/vt.py", line 32, in <module>
    from ..loader import VirtTestLoader
  File "/home/kar/workspace/avocado-vt/avocado_vt/loader.py", line 29, in <module>
    from virttest import storage
  File "/home/kar/workspace/avocado-vt/virttest/storage.py", line 22, in <module>
    from virttest import nbd
  File "/home/kar/workspace/avocado-vt/virttest/nbd.py", line 5, in <module>
    from virttest import utils_misc
  File "/home/kar/workspace/avocado-vt/virttest/utils_misc.py", line 83, in <module>
    from virttest import utils_disk
  File "/home/kar/workspace/avocado-vt/virttest/utils_disk.py", line 27, in <module>
    from virttest import remote
  File "/home/kar/workspace/avocado-vt/virttest/remote.py", line 13, in <module>
    from aexpect.remote import *

usage: avocado run [-h] [-p NAME_VALUE] [--test-runner RUN.TEST_RUNNER] [-d]
......
avocado run: error: unrecognized arguments: --vt-config
Perhaps a plugin is missing; run 'avocado plugins' to list the installed ones
`

**Comment by zhenyzha** [comment]
> Hi @zhenyzha,
> 
> > @pevogam hello This patch causes the following error to be reported when tp-qemu is running:
> > ```
> > avocado run: error: unrecognized arguments: --vt-config
> > Perhaps a plugin is missing; run 'avocado plugins' to list the installed ones
> > ```
> 
> This is odd as I wouldn't expect arguments on the command line to be affected in any way. Do you have a link to the CI logs or other output logs?

Hi @pevogam

Found the reason, because we did not update the aexpect version.
We will repair it internally, thanks for the feedback.

**Pull request merged**
